core/app_queue.go

```diff
@@ -0,0 +1,261 @@
+// Copyright 2022 Board of Trustees of the University of Illinois.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"fmt"
+	"notifications/core/model"
+	"notifications/driven/storage"
+	"time"
+
+	"github.com/rokwire/logging-library-go/v2/logs"
+)
+
+type queueLogic struct {
+	logger *logs.Logger
+
+	storage  Storage
+	firebase Firebase
+
+	//timer
+	queueTimer *time.Timer
+	timerDone  chan bool
+}
+
+func (q queueLogic) start() {
+	q.logger.Info("queueLogic start")
+
+	q.processQueue()
```

Comment:
I think there may be an issue here if the queue is locked when the instance first starts. In this case, it looks to me like the instance will never attempt to process the queue unless the specific instance receives a new message through the API. In normal functioning I think that is ok, but if some of the instances crash I think we could end up in a state where no instances are actively processing the timed items in the queue. One potential solution could be to periodically retry processing the queue until it is able to acquire the lock.